cekit/descriptor/image.py

```diff
@@ -45,48 +50,318 @@ def __init__(self, descriptor, artifact_dir):
+            self._descriptor['ports'] = ports.values()
+
+            module_repositories = self._to_dict(self.modules.repositories)
+            for repository in override.modules.repositories:
+                name = repository.name
+                if name in module_repositories:
+                    module_repositories[name] = repository.merge(module_repositories[name])
+                else:
+                    module_repositories[name] = repository
+            self.modules._descriptor['repositories'] = module_repositories.values()
+
+            package_repositories = self._to_dict(self.packages.repositories)
+            for repository in override.packages.repositories:
+                name = repository.name
+                if name in package_repositories:
+                    package_repositories[name] = repository.merge(package_repositories[name])
+                else:
+                    package_repositories[name] = repository
+            self.packages._descriptor['repositories'] = package_repositories.values()
+
+            if override.osbs != None:
+                self.osbs = override.osbs.merge(self.osbs)
+
+            # should we really allow overrides to contribute content?
+            for package in override.packages.install:
+                if package not in self.packages.install:
+                    self.packages.install.append(package)
+
+            artifact_overrides = self._image_overrides['artifacts']
+            image_artifacts = self._to_dict(self.artifacts)
+            for artifact in override.artifacts:
+                name = artifact.name
+                # collect override so we can apply it to modules
+                artifact_overrides[name] = artifact
+                # add it to the list of everything
+                self._all_artifacts[name] = artifact
+                if name in image_artifacts:
+                    # apply override to image descriptor
+                    image_artifacts[name] = artifact
+            self._descriptor['artifacts'] = image_artifacts.values()
+
+            module_overrides = self._image_overrides['modules']
+            image_modules = self._to_dict(self.modules.install)
+            for module in override.modules.install:
+                name = module.name
+                # collect override so we can apply it to modules.
+                # this allows us to override module versions without affecting ordering.
+                module_overrides[name] = module
+                if name in image_modules:
+                    # apply override to image descriptor
+                    image_modules[name] = module
+            self.modules._descriptor['install'] = image_modules.values()
+
+            if override.run != None:
+                if self.run:
+                    self.run = override.run.merge(self.run)
+                else:
+                    self.run = override.run
+
+    def apply_module_overrides(self, module_registry):
```

Comment:
its merging logic duplicated again. If its need to be tweaked why not to override merge() method in Overrides/Modules descriptors?

Reply:
See comment above.  No duplication, different behavior.  Then only thing changed in the modules are the artifact definitions.  The rest of this just collects the required modules, package repositories, and the result of all run specifications in the modules.

At the end of this processing, the image object has all of its fields updated from overrides, as well as the result of merging the `run` field from all overrides and modules, and a single list of modules to be installed.  The template then proceeds to apply each module, in order, to the image (i.e. we're not merging all the content from all modules into a single descriptor anymore).